File: compositor/src/main/java/org/westford/compositor/core/SurfaceView.java
package org.westford.compositor.core;

import com.google.auto.factory.AutoFactory;
import com.google.auto.factory.Provided;
import org.freedesktop.wayland.server.WlSurfaceResource;
import org.westford.Signal;
import org.westford.Slot;
import org.westford.compositor.core.calc.Mat4;
import org.westford.compositor.core.calc.Vec4;
import org.westford.compositor.protocol.WlSurface;

import javax.annotation.Nonnull;
import java.util.Optional;

@AutoFactory(allowSubclasses = true,
             className = "PrivateSurfaceViewFactory")
public class SurfaceView {

    @Nonnull
    private final Signal<SurfaceView, Slot<SurfaceView>> destroyedSignal = new Signal<>();
    @Nonnull
    private final Signal<Point, Slot<Point>>             positionSignal  = new Signal<>();

    @Nonnull
    private Optional<SurfaceView> parent = Optional.empty();

    @Nonnull
    private final Compositor        compositor;
    @Nonnull
    private final WlSurfaceResource wlSurfaceResource;

    @Nonnull
    private Mat4 positionTransform;
    @Nonnull
    private Mat4 transform;
    @Nonnull
    private Mat4 inverseTransform;

    private boolean enabled  = true;
    private boolean drawable = false;

    SurfaceView(@Provided @Nonnull final Compositor compositor,
                @Nonnull WlSurfaceResource wlSurfaceResource,
                @Nonnull Mat4 positionTransform,
                @Nonnull Mat4 transform,
                @Nonnull Mat4 inverseTransform) {
        this.compositor = compositor;
        this.wlSurfaceResource = wlSurfaceResource;
        this.positionTransform = positionTransform;
        this.transform = transform;
        this.inverseTransform = inverseTransform;
    }

    @Nonnull
    public WlSurfaceResource getWlSurfaceResource() {
        return this.wlSurfaceResource;
    }

    @Nonnull
    public Mat4 getPositionTransform() {
        return this.positionTransform;
    }

    private void setPosition(final Mat4 positionTransform) {
        this.positionTransform = positionTransform;

        final WlSurface wlSurface        = (WlSurface) getWlSurfaceResource().getImplementation();
        final Surface   surface          = wlSurface.getSurface();
        final Mat4      surfaceTransform = surface.getTransform();

        this.transform = this.positionTransform.multiply(surfaceTransform);
        this.inverseTransform = this.transform.invert();
    }

    public void setPosition(@Nonnull final Point global) {
        setPosition(Transforms.TRANSLATE(global.getX(),
                                         global.getY()));
        getPositionSignal().emit(global);

        this.compositor.requestRender();
    }

    @Nonnull
    public Signal<Point, Slot<Point>> getPositionSignal() {
        return this.positionSignal;
    }

    public void onApply(@Nonnull final SurfaceState surfaceState) {

        this.drawable = surfaceState.getBuffer()
                                    .isPresent();

        final Point deltaPosition = surfaceState.getDeltaPosition();
        final int   dx            = deltaPosition.getX();
        final int   dy            = deltaPosition.getY();

        setPosition(this.positionTransform.multiply(Transforms.TRANSLATE(dx,
                                                                         dy)));
    }

    @Nonnull
    public Signal<SurfaceView, Slot<SurfaceView>> getDestroyedSignal() {
        return this.destroyedSignal;
    }

    public void destroy() {
        final WlSurface wlSurface = (WlSurface) this.wlSurfaceResource.getImplementation();
        final Surface   surface   = wlSurface.getSurface();
        surface.getViews()
               .remove(this);

        this.destroyedSignal.emit(this);
        removeParent();
    }

    @Nonnull
    public Optional<SurfaceView> getParent() {
        return this.parent;
    }

    public void setParent(@Nonnull final SurfaceView parent) {
        removeParent();
        parent.getDestroyedSignal()
              .connect(event -> destroy());
        this.parent = Optional.of(parent);
    }

    public void removeParent() {
        this.parent.ifPresent(parentSurfaceView -> this.parent = Optional.empty());
    }

    /**
     * Conveniently translate from a compositor global coordinate to a view local coordinate.
     *
     * @param global A point from the compositor global plane.
     *
     * @return A point in view local plane.
     */
    @Nonnull
    public Point local(@Nonnull final Point global) {
        final Vec4 localPoint = this.inverseTransform.multiply(global.toVec4());
        return Point.create((int) localPoint.getX(),
                            (int) localPoint.getY());
    }

    /**
     * Conveniently translate from a view local coordinate to a compositor global coordinate.
     *
     * @param surfaceLocal A point from the view local plane.
     *
     * @return A point in the compositor global plane.
     */
    @Nonnull
    public Point global(@Nonnull final Point surfaceLocal) {
        final Vec4 globalPoint = this.transform.multiply(surfaceLocal.toVec4());
        return Point.create((int) globalPoint.getX(),
                            (int) globalPoint.getY());
    }

    /**
     * Indicates if this view should be rendered.
     *
     * @return
     */
    public boolean isEnabled() {
        return this.enabled;
    }

    public void setEnabled(final boolean enabled) {
        this.enabled = enabled;
    }

    /**
     * Indicates if this view is drawable ie. does it have a buffer that can be rendered.
     *
     * @return
     */
    public boolean isDrawable() {
        return this.drawable;
    }

    /**
     * Inverse of {{@link #getTransform()}}. Translates from compositor global coordinates to view local coordinates.
     *
     * @return
     */
    @Nonnull
    public Mat4 getInverseTransform() {
        return inverseTransform;
    }

    /**
     * Contains all view specific transformations, this includes positioning, rotation etc. of the view.
     * <p>
     * Translates from view local coordinates to compositor global coordinates.
     *
     * @return
     */
    @Nonnull
    public Mat4 getTransform() {
        return transform;
    }
}

File: compositor/src/main/java/org/westford/compositor/core/SurfaceView.kt
package org.westford.compositor.core

import com.google.auto.factory.AutoFactory
import com.google.auto.factory.Provided
import org.freedesktop.wayland.server.WlSurfaceResource
import org.westford.Signal
import org.westford.Slot
import org.westford.compositor.core.calc.Mat4
import org.westford.compositor.core.calc.Vec4
import org.westford.compositor.protocol.WlSurface
import java.util.Optional

@AutoFactory(allowSubclasses = true, className = "PrivateSurfaceViewFactory")
class SurfaceView internal constructor(@param:Provided private val compositor: Compositor,
                                       val wlSurfaceResource: WlSurfaceResource,
                                       positionTransform: Mat4,
                                       transform: Mat4,
                                       inverseTransform: Mat4) {

    val destroyedSignal = Signal<SurfaceView, Slot<SurfaceView>>()
    val positionSignal = Signal<Point, Slot<Point>>()

    var parent = Optional.empty<SurfaceView>()
        private set

    var positionTransform: Mat4
        private set
    /**
     * Contains all view specific transformations, this includes positioning, rotation etc. of the view.
     *
     *
     * Translates from view local coordinates to compositor global coordinates.

     * @return
     */
    var transform: Mat4
        private set
    /**
     * Inverse of {[.getTransform]}. Translates from compositor global coordinates to view local coordinates.

     * @return
     */
    var inverseTransform: Mat4
        private set

    /**
     * Indicates if this view should be rendered.

     * @return
     */
    var isEnabled = true
    /**
     * Indicates if this view is drawable ie. does it have a buffer that can be rendered.

     * @return
     */
    var isDrawable = false
        private set

    init {
        this.positionTransform = positionTransform
        this.transform = transform
        this.inverseTransform = inverseTransform
    }

    private fun setPosition(positionTransform: Mat4) {
        this.positionTransform = positionTransform

        val wlSurface = wlSurfaceResource.implementation as WlSurface
        val surface = wlSurface.surface
        val surfaceTransform = surface.transform

        this.transform = this.positionTransform.multiply(surfaceTransform)
        this.inverseTransform = this.transform.invert()
    }

    fun setPosition(global: Point) {
        setPosition(Transforms.TRANSLATE(global.x,
                global.y))
        positionSignal.emit(global)

        this.compositor.requestRender()
    }

    fun onApply(surfaceState: SurfaceState) {

        this.isDrawable = surfaceState.buffer
                .isPresent

        val deltaPosition = surfaceState.deltaPosition
        val dx = deltaPosition.x
        val dy = deltaPosition.y

        setPosition(this.positionTransform.multiply(Transforms.TRANSLATE(dx,
                dy)))
    }

    fun destroy() {
        val wlSurface = this.wlSurfaceResource.implementation as WlSurface
        val surface = wlSurface.surface
        surface.views
                .remove(this)

        this.destroyedSignal.emit(this)
        removeParent()
    }

    fun setParent(parent: SurfaceView) {
        removeParent()
        parent.destroyedSignal
                .connect({ event -> destroy() })
        this.parent = Optional.of(parent)
    }

    fun removeParent() {
        this.parent.ifPresent { parentSurfaceView -> this.parent = Optional.empty<SurfaceView>() }
    }

    /**
     * Conveniently translate from a compositor global coordinate to a view local coordinate.

     * @param global A point from the compositor global plane.
     * *
     * *
     * @return A point in view local plane.
     */
    fun local(global: Point): Point {
        val localPoint = this.inverseTransform.multiply(global.toVec4())
        return Point.create(localPoint.x.toInt(),
                localPoint.y.toInt())
    }

    /**
     * Conveniently translate from a view local coordinate to a compositor global coordinate.

     * @param surfaceLocal A point from the view local plane.
     * *
     * *
     * @return A point in the compositor global plane.
     */
    fun global(surfaceLocal: Point): Point {
        val globalPoint = this.transform.multiply(surfaceLocal.toVec4())
        return Point.create(globalPoint.x.toInt(),
                globalPoint.y.toInt())
    }
}
